// homey-meshdriver/lib/zigbee/system/capabilities/windowcoverings_set/closuresWindowCovering.js
'use strict';

const util = require('./../../../../util');

const UP_OPEN = 'upOpen';
const DOWN_CLOSE = 'downClose';
const CLUSTER_CLOSURES_WINDOW_COVERING = 'closuresWindowCovering';

module.exports = {
  set: 'goToLiftPercentage',
  async setParser(value) {

    // This boolean can be used to determine if incoming report are generated by a set command by Homey or externally
    this._reportDebounceEnabled = true;

    // Override goToLiftPercentage to enforce blind to open/close completely
    if (value === 0 || value === 1) {
      this.log(`do ${value === 1 ? UP_OPEN : DOWN_CLOSE}`);
      await this.node.endpoints[this.getClusterEndpoint(CLUSTER_CLOSURES_WINDOW_COVERING)].clusters.closuresWindowCovering.do(value === 1 ? UP_OPEN : DOWN_CLOSE, []);
      await this.setCapabilityValue('windowcoverings_set', value);
      return null;
    }

    // Send goToLiftPercentage command
    const mappedValue = util.mapValueRange(0, 1, 100, 0, value);
    return {
      percentageliftvalue: Math.round(mappedValue), // Round, otherwise might not be accepted by device
    };
  },
  get: 'currentPositionLiftPercentage',
  reportParser(value) {
    if (value => 0 && value <= 100) {
      return util.mapValueRange(0, 100, 1, 0, value);
    }
    return null;
  },
  report: 'currentPositionLiftPercentage',
  getOpts: {
    getOnStart: true,
  },
};
