// homey-meshdriver/lib/zwave/system/capabilities/alarm_tamper/NOTIFICATION.js
'use strict';

module.exports = {
	get: 'NOTIFICATION_GET',
	getOpts: {
		getOnOnline: true,
	},
	getParser: () => ({
		'V1 Alarm Type': 0,
		Event: 3,
		'Notification Type': 'Home Security',
	}),
	report: 'NOTIFICATION_REPORT',
	reportParser: report => {
		if ((report && report['Notification Type'] === 'Home Security' ||
												report['Notification Type'] === 'Burglar') &&
												report.hasOwnProperty('Event (Parsed)')) {

			if (report['Event (Parsed)'] === 'Tampering, Product covering removed' ||
																report['Event (Parsed)'] === 'Tampering, Invalid Code' ||
																report['Event (Parsed)'] === 'Tampering, Product Moved') {
				return true;
			}
			if (report['Event (Parsed)'] === 'Event inactive' && (!report.hasOwnProperty('Event Parameter') ||
																typeof report['Event Parameter'][0] === 'undefined' ||
																report['Event Parameter'][0] === 3 ||
																report['Event Parameter'][0] === 4 ||
																report['Event Parameter'][0] === 9)) {
				return false;
			}
		}
		return null;
	},
};
